### PR TITLE
Add `prefer-default-parameters` rule

### LINKED ISSUE #29: Prefer default parameters
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/29)

Applies only to `es6` mode:

Bad:

``` js
function (foo) {
  foo = foo || 'bar';
  // ...
}
```

Good:

``` js
function (foo = 'bar') {
  // ...
}
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)

### Submitted pull Requests
- [#632 Add `prefer-default-parameters` rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/632)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by medusalix**
Some notes: I've tried to keep the rule simple by only reporting `literal` assignments (see [this](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/29#issuecomment-502794939) comment). Combining parameters with e.g. a return value of another function or variable might be desired in some cases and should in my opinion not be reported. I've had to work around arrow functions with only one parameter and no parentheses (`foo => {}`) to make sure the default parameter gets inserted correctly by including the missing parentheses.

Fixes #29

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#29: Prefer default parameters](https://issuehunt.io/repos/55832243/issues/29)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: rules/prefer-default-parameters.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+	return {
+		FunctionDeclaration: node => {
+			currentFunction = node;
+		},
+		ArrowFunctionExpression: node => {
+			currentFunction = node;
+		},
+		'FunctionDeclaration:exit': () => {
+			currentFunction = undefined;
```
Comment (fisker) [inline rules/prefer-default-parameters.js:112]: I think this is not right,  if assign is after a  inner function, should still in the outer function.

File: rules/prefer-default-parameters.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+const declarationSelector = [
+	'VariableDeclaration',
+	'[declarations.0.type="VariableDeclarator"]'
+].join('');
+
+const isDefaultExpression = (left, right) =>
+	left &&
+	right &&
+	left.type === 'Identifier' &&
+	right.type === 'LogicalExpression' &&
```
Comment (fisker) [inline rules/prefer-default-parameters.js:20]: you  forgot  check  operator

File: rules/prefer-default-parameters.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+			left: {name: secondId},
+			right: {raw: literal}
+		} = right;
+
+		// Check if literal is assigned to the same identifier
+		if (assignment && firstId !== secondId) {
+			return;
+		}
+
+		const parameter = currentFunction.params.find(parameter =>
```
Comment (fisker) [inline rules/prefer-default-parameters.js:81]: I think we only need check the laster parameter.

File: test/prefer-default-parameters.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		},
+		{
+			code: 'function abc(foo) { foo = foo || true; }',
+			output: 'function abc(foo = true) { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo, bar) { foo = foo || \'bar\'; baz(); }',
+			output: 'function abc(foo = \'bar\', bar) { baz(); }',
+			errors: [error]
```
Comment (fisker) [inline test/prefer-default-parameters.js:72]: This is why I said, we shoul only check last parameter, not good code.

File: test/prefer-default-parameters.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		},
+		{
+			code: 'function abc(foo) { const bar = foo || \'bar\'; }',
+			output: 'function abc(bar = \'bar\') { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { let bar = foo || \'bar\'; }',
+			output: 'function abc(bar = \'bar\') { }',
```
Comment (fisker) [inline test/prefer-default-parameters.js:81]: Not sure about this `function abc(foo = \'bar\') { const bar = foo;}`, this should easier, but you already did better. Any chance this would break something?

**Comment by fisker** [comment]
One more thing, since the fix  I not  safe , we should use suggestion api, but don't worry, we can easily convert it.  [example](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/prefer-number-properties.js]

File: test/prefer-default-parameters.js (hunk 4 above)
Comment (medusalix) [reply test/prefer-default-parameters.js:73]: Why do you think only the last parameter should be checked?

**Comment by medusalix** [comment]
The fix should be safe because I only handle the cases where a literal is used as a default parameter. See [this](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/29#issuecomment-502794939) comment.

**Comment by medusalix** [comment]
> Can you update code with code in function body to multiple lines?

Do you mean the test cases?

File: test/prefer-default-parameters.js (hunk 5 above)
Comment (medusalix) [reply test/prefer-default-parameters.js:82]: Yeah, if some code relies on the presence of the `foo` parameter it might berak. I'm going to use the suggestions API for this.

**Comment by fisker** [comment]
```js
'function abc() { let foo = 123; const bar = foo || \'bar\' }',
```

to

```js
`
function abc() {
  let foo = 123;
  const bar = foo || 'bar' 
}
`,
```

File: test/prefer-default-parameters.js (hunk 5 above)
Comment (fisker) [reply test/prefer-default-parameters.js:82]: We can check if `foo` is used. If not used, in the way you already did, otherwise fix to the code I wrote.

WDYT?

File: test/prefer-default-parameters.js (hunk 4 above)
Comment (fisker) [reply test/prefer-default-parameters.js:73]: See https://eslint.org/docs/rules/default-param-last

File: rules/prefer-default-parameters.js (hunk 1 above)
Comment (medusalix) [reply rules/prefer-default-parameters.js]: Fixed using a function stack in the latest commit.

**Comment by medusalix** [comment]
> I think we only need check the last parameter.

We still need to *check* all of them to see if it's actually one of the parameters, but I chose to skip the fixing if it's not the last parameter.

File: rules/prefer-default-parameters.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		},
+		ArrowFunctionExpression: node => {
+			functionStack.push(node);
+		},
+		'FunctionDeclaration:exit': () => {
+			functionStack.pop();
+		},
+		'ArrowFunctionExpression:exit': () => {
+			functionStack.pop();
+		},
```
Comment (fisker) [inline rules/prefer-default-parameters.js:134]: I think `:function` and `::function:exit` should work.

**Comment by fisker** [comment]
> We still need to check all of them to see if it's actually one of the parameters, but I chose to skip the fixing if it's not the last parameter.

I think we can simple this logic, only report one problem each time on one function. but we need consider following cases

`function foo(a, b, c=1){}` check `b` not `c`, 
`function foo(a, b, ...c){}` also check `b`.

**Comment by fisker** [comment]
Another edge case, since the assignment can be used as value.

```js
function foo(a){
 doSomething(
   a = a || 1
)

doAnotherThing(a)
}
```

or

```js
function foo(a){
 const b = a = a || 1
}
```

So, we should choose leave `a` there or ignore those case.

My suggestion, alway leave `a` there unless parent is `ExpressionStatement`.

**Comment by fisker** [comment]
@medusalix What do  you think the suggestion above?

**Comment by medusalix** [comment]
Sorry for the late reply, I'm pretty busy right now, I'll take a look at your suggestions when I get the time.

**Comment by medusalix** [comment]
> Another edge case, since the assignment can be used as value.

Yeah, that's already checked by the selector: `ExpressionStatement[expression.type="AssignmentExpression"]`. I've included both of your tests.

File: test/prefer-default-parameters.js [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+				}
+			`]
+		}),
+		invalidTestCase({
+			code: outdent`
+				function abc(foo, bar) {
+					foo = foo || 'bar';
+					baz();
+				}
+			`
```
Comment (fisker) [inline test/prefer-default-parameters.js:158]: I thought we agree only check the last parameter?

Comment (medusalix) [reply test/prefer-default-parameters.js]: Yes, sorry, I assumed we would still report them. I fixed it in the last commit.

**Comment by fisker** [comment]
1. Please fix tests I added.
2. `??` operator is supported by eslint now, please check it too.

File: rules/prefer-default-parameters.js [hunk 8]
```diff
@@ -115,6 +121,14 @@ const create = context => {
 			functionStack.push(node);
 		},
 		':function:exit': () => {
+			for (const [id, report] of reportedIdentifiers) {
```
Comment (fisker) [inline rules/prefer-default-parameters.js:124]: @medusalix 

We use `variable` to analysis the identifiers https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/prefer-set-has.js#L143.

Comment (medusalix) [reply rules/prefer-default-parameters.js:124]: Ah, even better. No need to reinvent the wheel I guess.

Comment (fisker) [reply rules/prefer-default-parameters.js:124]: I'm not sure if we can check this,

```js
			function abc(foo) {
				baz();
				foo = foo || 'bar';
				function baz() {
					console.log(foo);
				}
			}
```

In this case, `foo` is used before the assignment, but how to know?

Comment (medusalix) [reply rules/prefer-default-parameters.js:124]: Yeah, I don't think we can check whether certain functions have side-effects. But we use the suggestions API, so users still have the option to ignore the fixes.

**Comment by medusalix** [comment]
I moved one of your test cases because it can be fixed. References are now checked using `findVariable`.

**Comment by fisker** [comment]
> I moved one of your test cases because it can actually be fixed. References are now checked using `findVariable`.

Good, you checked the `foo`

File: rules/prefer-default-parameters.js (hunk 8 above)
Comment (fisker) [reply rules/prefer-default-parameters.js]: I don't know, though users can ignore the fix, but it will still reported.

This seems the tricky part.

Comment (fisker) [reply rules/prefer-default-parameters.js]: `eslint-utils` has a function called [`hasSideEffect`](https://eslint-utils.mysticatea.dev/api/ast-utils.html#hassideeffect), maybe we can check all nodes before the assignment, ignore those cases hasSideEffect?

Comment (medusalix) [reply rules/prefer-default-parameters.js]: That works, but it also ignores this (and 5 other test cases):

```js
function abc(bar) {
	foo();
	bar = bar || 123;
}
```

It's only a heuristic, so we get lots of valid test cases that would otherwise be considered invalid.

Comment (fisker) [reply rules/prefer-default-parameters.js]: @sindresorhus What do you think?

Comment (sindresorhus) [reply rules/prefer-default-parameters.js]: I'm fine with ignoring such cases. That code should use default parameters anyway.

File: rules/prefer-default-parameters.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+			right: {raw: literal}
+		} = right;
+
+		// Check if literal is assigned to the same identifier
+		if (assignment && firstId !== secondId) {
+			return;
+		}
+
+		const variable = findVariable(context.getScope(), secondId);
+		const maxReferences = assignment ? 2 : 1;
```
Comment (fisker) [inline rules/prefer-default-parameters.js:87]: I feel this not good, but I don't know what to do

```js
			function abc(foo = 123) {
				foo = foo || 'bar';
				console.log(foo)
			}
```

**Comment by fisker** [comment]
I think the problem we can't resolve here is

1.

```js
function abc(foo) {
	foo = foo || 'bar';
	console.log(foo);
}
```

VS


```js
function abc(foo) {
	console.log(foo);
	foo = foo || 'bar';
}
```

2.

```js
function abc(foo) {
	useFoo();
	foo = foo || 'bar';

	function useFoo() {
		console.log(foo);
	}
}
```

VS

```js
function abc(foo) {
	notUseFoo();
	foo = foo || 'bar';

	function notUseFoo() {
	}
}
```

**Comment by sindresorhus** [comment]
Bump

**Comment by medusalix** [comment]
I think I've found a pretty good solution for the first point @fisker outlined above:

1. If the parameter is being reassigned, make sure that there are no references to the parameter before the default-assignment:

```js
foo = foo || 123; // Reference 1 and 2
console.log(foo); // Reference 3
// First reference is the default-assignment; invalid, can be fixed.
```

```js
console.log(foo); // Reference 1
foo = foo || 123; // Reference 2 and 3
// First reference is the console.log statement; valid.
```

2. If the parameter is default-assigned to a new variable, make sure that there are no additional references to the old parameter *at all*:

```js
const bar = foo || 'bar'; // Reference 1
console.log(foo, bar);    // Reference 2
// More than one reference; valid.
```

```js
const bar = foo || 'bar'; // Reference 1
console.log(bar);
// Only one reference; invalid, can be fixed.
```

**Comment by medusalix** [comment]
I've done some tests with the `hasSideEffect` function. I believe implementing side-effect checking defeats the purpose of this rule. Consider the following example:

```js
function abc(foo) {
	let bar = true;
	bar = false;

	foo = foo || 123;
	console.log(foo);
}
```

This example would be ignored by the rule because `hasSideEffect` returns `true` for the `bar = false` statement, even though it doesn't have any effects on the `foo` parameter *at all*.

@fisker I think ignoring cases where there is any `CallExpression` before the default-assignment would be a better solution. What do you think?

**Comment by fisker** [comment]
I'm running out of ideas. :smile:

File: rules/prefer-default-parameters.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+				fix: fixer => [
+					fixer.replaceText(parameter, replacement),
+					fixDefaultExpression(fixer, source, node)
+				]
+			}]
+		});
+	};
+
+	return {
+		':function': node => {
```
Comment (fisker) [inline rules/prefer-default-parameters.js:127]: Do you think it should work on `ObjectMethod`? 

```js
foo = {
  method(foo) {
    foo = foo || 123;
  }
};
```

and `ClassMethod`

```js
class Foo {
  method(foo) {
    foo = foo || 123;
  }
}
```

File: test/prefer-default-parameters.js [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+						bar = bar || 'foo';
+					}
+					function ghi(bay = 'bar') {
+					}
+					foo = foo || 'bar';
+				}
+			`]
+		})
+	]
+});
```
Comment (fisker) [inline test/prefer-default-parameters.js:403]: `FunctionExpression` seems not tested

```js
foo = function (foo){
  foo = foo || 123;
}
```

**Review by fisker — CHANGES_REQUESTED** [review]

File: rules/prefer-default-parameters.js (hunk 10 above)
Comment (medusalix) [reply rules/prefer-default-parameters.js:181]: Yes, I think it should. I added some tests.

File: test/prefer-default-parameters.js (hunk 11 above)
Comment (medusalix) [reply test/prefer-default-parameters.js:657]: I added some tests.

**Comment by medusalix** [comment]
@fisker I've implemented the `CallExpression` check I described in my last comment and added lots of tests for it.

File: rules/prefer-default-parameters.js [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+			}
+		} else if (containsCallExpression(source, value)) {
+			return true;
+		}
+	}
+
+	return false;
+};
+
+const hasSideEffects = (source, function_, node) => {
```
Comment (sindresorhus) [inline rules/prefer-default-parameters.js:53]: https://eslint-utils.mysticatea.dev/api/ast-utils.html#hassideeffect ?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
// @fisker

**Review by fisker — APPROVED** [review]
I forgot the review progress, I read all comments from beginning again, seems good enough, we can still improve it if any issue found.
